File: ts/src/inserirPaleta.ts
import fs from 'fs'
import { Failure, Result, Success, failure, result, success } from './result'
import { makeRail } from './railway'
import { ErroInsercaoPaleta, mensagemErroInsercaoPaleta } from './erros'
import { InserirPaletaDto } from './dtos'
import { CorPaleta, Paleta } from './paleta'
import { Repositorio } from './repositorio'

const lerArquivoDePaleta = (nomeDoArquivo: string): Result<string, ErroInsercaoPaleta> => {
  try {
    return success(fs.readFileSync(nomeDoArquivo).toString())
  } catch (error) {
    return failure({
      kind: 'erro_ao_ler_arquivo',
      nomeDoArquivo,
      mensagemDeErro: (error as Error).message
    })
  }
}

const separarArquivoEmLinhas = (arquivo: string): string[] =>
  arquivo.split('\n')

const converterLinhasEmPaleta = (linhas: string[]): Result<InserirPaletaDto[], ErroInsercaoPaleta> => {
  const dadosPaleta =
    linhas
      .map(linha => linha
        .split(' ')
        .map(l => l.trim())
        .filter(l => l !== ''))
      .map((linha, indice): Result<InserirPaletaDto, [number, string]> => {
        if (linha.length !== 4) return failure([indice, linha.join(' ')])

        let [r, g, b, nome] = linha

        return success({ rgb: [r, g, b], nome })
      })

  return result.elevate((linhas) => ({ kind: 'formato_de_paleta_invalido', linhas }), dadosPaleta)
}

const converterRgbEmCores = (dadosPaleta: InserirPaletaDto[]): Result<CorPaleta[], ErroInsercaoPaleta> => {
  const cores =
    dadosPaleta.map((dadosCor, indice): Result<CorPaleta, [number, string]> => {
      const [rString, gString, bString] = dadosCor.rgb

      const r = Number(rString)
      const g = Number(gString)
      const b = Number(bString)

      const valorDeCorInvalido = (cor: number) => isNaN(cor) || cor > 255 || cor < 0

      if ([r, g, b].findIndex(valorDeCorInvalido) !== -1) {
        return failure([indice, dadosCor.rgb.join(', ')])
      }

      const toHex = (number: number) => number.toString(16).padStart(2, '0')
      const cor = `#${toHex(r)}${toHex(g)}${toHex(b)}`

      return success({ nome: dadosCor.nome, cor })
    })

  return result.elevate((linhas) => ({ kind: 'valor_de_cor_invalido', linhas }), cores)
}

const criarPaleta = (nome: string, cores: CorPaleta[]): Paleta => ({
  nome,
  cores
})

const salvarPaletaNoBancoDeDados = (repositorio: Repositorio, paleta: Paleta): void => {
  repositorio.inserirPaleta(paleta)
    .catch((e: Error) => {
      throw new Error(e.message)
    })
}

const imprimirResultado = (resultado: Result<Paleta, ErroInsercaoPaleta>) => {
  if (result.isSuccess(resultado)) {
    const paleta = (resultado as Success<Paleta>).success
    console.log(`Paleta ${paleta.nome} inserida com sucesso.\n`)
  } else {
    const erro = (resultado as Failure<ErroInsercaoPaleta>).failure
    console.log(mensagemErroInsercaoPaleta(erro))
  }
}

const gravarErro = (erro: ErroInsercaoPaleta) => {
  console.error(erro)
}

export const inserirPaleta = (repositorio: Repositorio, nomeDaPaleta: string, nomeDoArquivo: string) => {
  makeRail(lerArquivoDePaleta)
    .map(separarArquivoEmLinhas)
    .then(converterLinhasEmPaleta)
    .then(converterRgbEmCores)
    .map((cores) => criarPaleta(nomeDaPaleta, cores))
    .tee((paleta) => salvarPaletaNoBancoDeDados(repositorio, paleta))
    .teeError(gravarErro)
    .finally(imprimirResultado)
    .run(nomeDoArquivo)
}

File: POF/InserirPaleta.fs
module InserirPaleta

open System
open System.IO
open System.Drawing
open Railway
open Paleta
open Repositorio
open Erros
open DTOs
open StringExtensions

let lerArquivoDePaleta nomeDoArquivo =
    try
        Ok <| File.ReadAllLines nomeDoArquivo
    with excecao ->
        Error(ErroAoLerArquivo(nomeDoArquivo, excecao.Message))

let converterLinhaEmPaleta linhas =
    let dividirLinha (str: string) =
        str
        |> String.split " "
        |> Array.map String.trimWhiteSpaces
        |> Array.filter String.isNotNullOrWhiteSpace

    let dadosPaletas =
        linhas
        |> Array.map dividirLinha
        |> Array.mapi (fun (index: int) (linha: string array) ->
            match linha with
            | [| r; g; b; nome |] -> Ok { RGB = r, g, b; Nome = nome }
            | _ -> Error(index, linha |> String.concat " "))

    if Array.exists Result.isError dadosPaletas then
        dadosPaletas
        |> Array.filter Result.isError
        |> Array.choose Result.toOptionError
        |> (FormatoDePaletaInvalido >> Error)

    else
        dadosPaletas |> Array.choose Result.toOption |> Ok

let validarRGB dadosPaletas =
    let validacoes =
        dadosPaletas
        |> Array.mapi (fun (index: int) (dadosPaleta: InserirPaletaDto) ->
            let r, g, b = dadosPaleta.RGB

            match [| r; g; b |] with
            | x when Array.exists (String.canBeConvertedToInt32 >> not) x -> Error(index, $"{r}, {g}, {b}")
            | _ -> Ok())

    if Array.exists Result.isError validacoes then
        validacoes
        |> Array.filter Result.isError
        |> Array.choose Result.toOptionError
        |> (FormatoDePaletaInvalido >> Error)
    else
        Ok dadosPaletas

let converterRGBEmCores dadosPaletas =
    let cores =
        dadosPaletas
        |> Array.mapi (fun (index: int) (dadosPaleta: InserirPaletaDto) ->
            let rString, gString, bString = dadosPaleta.RGB

            let r = Int32.Parse(rString)
            let g = Int32.Parse(gString)
            let b = Int32.Parse(bString)

            let valorDeCorInvalido cor = cor > 255 || cor < 0

            match [| r; g; b |] with
            | x when Array.exists valorDeCorInvalido x -> Error(index, $"{r}, {g}, {b}")
            | _ ->
                Ok
                    { Nome = dadosPaleta.Nome
                      Cor = Color.FromArgb(r, g, b) })

    if Array.exists Result.isError cores then
        cores
        |> Array.filter Result.isError
        |> Array.choose Result.toOptionError
        |> (ValorDeCorInvalido >> Error)
    else
        cores |> Array.choose Result.toOption |> Ok

let criarPaleta nomeDaPaleta cores = { Nome = nomeDaPaleta; Cores = cores }

let salvarPaletaNoBancoDeDados (repositorio: Repositorio) paleta = repositorio.InserirPaleta paleta

let registrarErro (erro: ErroInsercaoPaleta) = Console.Error.WriteLine(erro)

let finalizar =
    function
    | Error(e: ErroInsercaoPaleta) -> printfn "Erro ao executar: %s" (e.MensagemDeErro())
    | Ok(paleta: Paleta) -> printfn "A paleta %s foi incluída com sucesso" paleta.Nome

[<RequireQualifiedAccess>]
module InserirPaleta =
    let executar repositorio nomePaleta =
        lerArquivoDePaleta
        >> Result.bind converterLinhaEmPaleta
        >> Result.bind validarRGB
        >> Result.bind converterRGBEmCores
        >> Result.map (criarPaleta nomePaleta)
        >> Result.tee (salvarPaletaNoBancoDeDados repositorio >> Async.RunSynchronously)
        >> Result.teeError (registrarErro)
        >> finalizar
